File: frontend/src/types/command.ts
/**
 * Tipos TypeScript para o módulo de Comandas (estilo Trinks)
 * 
 * Representa o sistema de fechamento de conta com múltiplas formas de pagamento,
 * integrado com meios de pagamento e cálculo automático de taxas.
 */

export type CommandStatus = 'OPEN' | 'CLOSED' | 'CANCELED';

export type CommandItemType = 'SERVICO' | 'PRODUTO' | 'PACOTE';

/**
 * Comanda principal (conta do cliente)
 */
export interface Command {
  id: string;
  tenant_id: string;
  appointment_id?: string;
  customer_id: string;
  numero?: string;
  status: CommandStatus;
  
  // Valores financeiros
  subtotal: number;
  desconto: number;
  total: number;
  total_recebido: number;
  troco: number;
  saldo_devedor: number;
  
  // Opções de fechamento
  observacoes?: string;
  deixar_troco_gorjeta: boolean;
  deixar_saldo_divida: boolean;
  
  // Auditoria
  criado_em: string;
  atualizado_em: string;
  fechado_em?: string;
  fechado_por?: string;
  
  // Relacionamentos (eager loaded quando necessário)
  items?: CommandItem[];
  payments?: CommandPayment[];
}

/**
 * Item da comanda (serviço, produto, pacote)
 */
export interface CommandItem {
  id: string;
  command_id: string;
  tipo: CommandItemType;
  item_id: string; // ID do serviço/produto/pacote
  descricao: string;
  
  // Valores
  preco_unitario: number;
  quantidade: number;
  desconto_valor: number;
  desconto_percentual: number;
  preco_final: number;
  
  observacoes?: string;
  criado_em: string;
}

/**
 * Pagamento da comanda
 */
export interface CommandPayment {
  id: string;
  command_id: string;
  meio_pagamento_id: string;
  
  // Valores
  valor_recebido: number;
  taxa_percentual: number;
  taxa_fixa: number;
  valor_liquido: number;
  
  observacoes?: string;
  criado_em: string;
  criado_por?: string;
}

// ============================================================================
// DTOs para Requests
// ============================================================================

/**
 * Request para criar comanda
 */
export interface CreateCommandRequest {
  appointment_id?: string;
  customer_id: string;
  items: CommandItemInput[];
  observacoes?: string;
}

/**
 * Input de item ao criar comanda
 */
export interface CommandItemInput {
  tipo: CommandItemType;
  item_id: string;
  descricao: string;
  preco_unitario: string; // String conforme esperado pelo backend
  quantidade?: number;
  desconto_valor?: number;
  desconto_percentual?: number;
}

/**
 * Request para adicionar item à comanda existente
 */
export interface AddCommandItemRequest {
  tipo: CommandItemType;
  item_id: string;
  descricao: string;
  preco_unitario: string; // String conforme esperado pelo backend
  quantidade?: number;
  desconto_valor?: number;
  desconto_percentual?: number;
  observacoes?: string;
}

/**
 * Request para atualizar item
 */
export interface UpdateCommandItemRequest {
  preco_unitario?: string; // String conforme esperado pelo backend
  quantidade?: number;
  desconto_valor?: number;
  desconto_percentual?: number;
  observacoes?: string;
}

/**
 * Request para adicionar pagamento
 */
export interface AddCommandPaymentRequest {
  meio_pagamento_id: string;
  valor_recebido: string; // Backend espera string para valores monetários
  observacoes?: string;
}

/**
 * Request para fechar comanda
 */
export interface CloseCommandRequest {
  deixar_troco_gorjeta?: boolean;
  deixar_saldo_divida?: boolean;
  observacoes?: string;
}

// ============================================================================
// DTOs para Responses
// ============================================================================

export interface CommandResponse {
  id: string;
  tenant_id: string;
  appointment_id?: string;
  customer_id: string;
  numero?: string;
  status: CommandStatus;
  subtotal: string; // Money formatado
  desconto: string;
  total: string;
  total_recebido: string;
  troco: string;
  saldo_devedor: string;
  observacoes?: string;
  deixar_troco_gorjeta: boolean;
  deixar_saldo_divida: boolean;
  criado_em: string;
  atualizado_em: string;
  fechado_em?: string;
  fechado_por?: string;
  items?: CommandItemResponse[];
  payments?: CommandPaymentResponse[];
}

export interface CommandItemResponse {
  id: string;
  command_id: string;
  tipo: CommandItemType;
  item_id: string;
  descricao: string;
  preco_unitario: string;
  quantidade: number;
  desconto_valor: string;
  desconto_percentual: number;
  preco_final: string;
  observacoes?: string;
  criado_em: string;
}

export interface CommandPaymentResponse {
  id: string;
  command_id: string;
  meio_pagamento_id: string;
  valor_recebido: string;
  taxa_percentual: number;
  taxa_fixa: string;
  valor_liquido: string;
  observacoes?: string;
  criado_em: string;
  criado_por?: string;
}

// ============================================================================
// State Management Types
// ============================================================================

/**
 * Estado de um pagamento selecionado no form
 */
export interface SelectedPayment {
  meio_pagamento_id: string;
  nome: string;
  tipo: string;
  valor_recebido: number;
  taxa_percentual: number;
  taxa_fixa: number;
  valor_liquido: number;
  icone?: string;
  cor?: string;
}

/**
 * Estado do formulário de pagamento
 */
export interface PaymentFormState {
  selected_payments: SelectedPayment[];
  deixar_troco_gorjeta: boolean;
  deixar_saldo_divida: boolean;
  observacoes: string;
}

/**
 * Resumo financeiro calculado
 */
export interface ResumoFinanceiro {
  subtotal: number;
  desconto: number;
  total: number;
  total_recebido: number;
  total_taxas: number;
  total_liquido: number;
  falta: number;
  troco: number;
}

// ============================================================================
// Helpers
// ============================================================================

/**
 * Calcula valor líquido após aplicar taxas
 */
export function calcularValorLiquido(
  valorRecebido: number,
  taxaPercentual: number,
  taxaFixa: number
): number {
  const valorTaxaPercentual = (valorRecebido * taxaPercentual) / 100;
  return valorRecebido - valorTaxaPercentual - taxaFixa;
}

/**
 * Calcula resumo financeiro da comanda
 */
export function calcularResumoFinanceiro(
  subtotal: number,
  desconto: number,
  selectedPayments: SelectedPayment[]
): ResumoFinanceiro {
  const total = subtotal - desconto;
  const totalRecebido = selectedPayments.reduce((sum, p) => sum + p.valor_recebido, 0);
  const totalTaxas = selectedPayments.reduce(
    (sum, p) => sum + (p.valor_recebido - p.valor_liquido),
    0
  );
  const totalLiquido = selectedPayments.reduce((sum, p) => sum + p.valor_liquido, 0);
  const diferenca = totalRecebido - total;
  
  return {
    subtotal,
    desconto,
    total,
    total_recebido: totalRecebido,
    total_taxas: totalTaxas,
    total_liquido: totalLiquido,
    falta: diferenca < 0 ? Math.abs(diferenca) : 0,
    troco: diferenca > 0 ? diferenca : 0,
  };
}

/**
 * Valida se pode fechar a comanda
 */
export function canCloseCommand(
  resumo: ResumoFinanceiro,
  deixarSaldoDivida: boolean
): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  
  if (resumo.total_recebido === 0) {
    errors.push('Nenhum pagamento foi registrado');
  }
  
  if (resumo.falta > 0 && !deixarSaldoDivida) {
    errors.push(`Falta receber R$ ${resumo.falta.toFixed(2)}`);
  }
  
  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Formata valor monetário
 */
export function formatMoney(value: number | string): string {
  const num = typeof value === 'string' ? parseFloat(value) : value;
  return new Intl.NumberFormat('pt-BR', {
    style: 'currency',
    currency: 'BRL',
  }).format(num);
}

/**
 * Parse de string monetária para number
 */
export function parseMoney(value: string): number {
  return parseFloat(value.replace(/[^0-9,-]/g, '').replace(',', '.')) || 0;
}

File: backend/internal/domain/entity/command.go
package entity

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

// CommandStatus representa os possíveis status de uma comanda
type CommandStatus string

const (
	CommandStatusOpen     CommandStatus = "OPEN"
	CommandStatusClosed   CommandStatus = "CLOSED"
	CommandStatusCanceled CommandStatus = "CANCELED"
)

// Command representa uma comanda de atendimento
type Command struct {
	ID            uuid.UUID
	TenantID      uuid.UUID
	AppointmentID *uuid.UUID
	CustomerID    uuid.UUID
	Numero        *string
	Status        CommandStatus

	// Valores financeiros
	Subtotal      float64
	Desconto      float64
	Total         float64
	TotalRecebido float64
	Troco         float64
	SaldoDevedor  float64

	// Opções de fechamento
	Observacoes        *string
	DeixarTrocoGorjeta bool
	DeixarSaldoDivida  bool

	// Auditoria
	CriadoEm     time.Time
	AtualizadoEm time.Time
	FechadoEm    *time.Time
	FechadoPor   *uuid.UUID

	// Relacionamentos (carregados quando necessário)
	Items    []CommandItem
	Payments []CommandPayment
}

// NewCommand cria uma nova comanda
func NewCommand(tenantID, customerID uuid.UUID, appointmentID *uuid.UUID) (*Command, error) {
	if tenantID == uuid.Nil {
		return nil, errors.New("tenant_id é obrigatório")
	}
	if customerID == uuid.Nil {
		return nil, errors.New("customer_id é obrigatório")
	}

	now := time.Now()
	return &Command{
		ID:                 uuid.New(),
		TenantID:           tenantID,
		CustomerID:         customerID,
		AppointmentID:      appointmentID,
		Status:             CommandStatusOpen,
		Subtotal:           0,
		Desconto:           0,
		Total:              0,
		TotalRecebido:      0,
		Troco:              0,
		SaldoDevedor:       0,
		DeixarTrocoGorjeta: false,
		DeixarSaldoDivida:  false,
		CriadoEm:           now,
		AtualizadoEm:       now,
		Items:              []CommandItem{},
		Payments:           []CommandPayment{},
	}, nil
}

// CanClose verifica se a comanda pode ser fechada
func (c *Command) CanClose() error {
	if c.Status != CommandStatusOpen {
		return errors.New("comanda já está fechada ou cancelada")
	}

	if len(c.Items) == 0 {
		return errors.New("comanda sem itens não pode ser fechada")
	}

	// Se não permitir saldo devedor, deve estar totalmente pago
	if !c.DeixarSaldoDivida && c.TotalRecebido < c.Total {
		falta := c.Total - c.TotalRecebido
		return errors.New("falta receber R$ " + formatMoney(falta))
	}

	return nil
}

// Close fecha a comanda
func (c *Command) Close(userID uuid.UUID) error {
	if err := c.CanClose(); err != nil {
		return err
	}

	now := time.Now()
	c.Status = CommandStatusClosed
	c.FechadoEm = &now
	c.FechadoPor = &userID
	c.AtualizadoEm = now

	// Calcular troco ou saldo devedor
	c.CalculateBalance()

	return nil
}

// Cancel cancela a comanda
func (c *Command) Cancel() error {
	if c.Status != CommandStatusOpen {
		return errors.New("apenas comandas abertas podem ser canceladas")
	}

	c.Status = CommandStatusCanceled
	c.AtualizadoEm = time.Now()

	return nil
}

// AddItem adiciona um item à comanda
func (c *Command) AddItem(item CommandItem) error {
	if c.Status != CommandStatusOpen {
		return errors.New("não é possível adicionar itens a uma comanda fechada")
	}

	c.Items = append(c.Items, item)
	c.RecalculateTotals()

	return nil
}

// RemoveItem remove um item da comanda
func (c *Command) RemoveItem(itemID uuid.UUID) error {
	if c.Status != CommandStatusOpen {
		return errors.New("não é possível remover itens de uma comanda fechada")
	}

	for i, item := range c.Items {
		if item.ID == itemID {
			c.Items = append(c.Items[:i], c.Items[i+1:]...)
			c.RecalculateTotals()
			return nil
		}
	}

	return errors.New("item não encontrado")
}

// AddPayment adiciona um pagamento à comanda
func (c *Command) AddPayment(payment CommandPayment) error {
	if c.Status != CommandStatusOpen {
		return errors.New("não é possível adicionar pagamentos a uma comanda fechada")
	}

	c.Payments = append(c.Payments, payment)
	c.TotalRecebido += payment.ValorRecebido
	c.CalculateBalance()

	return nil
}

// RemovePayment remove um pagamento da comanda
func (c *Command) RemovePayment(paymentID uuid.UUID) error {
	if c.Status != CommandStatusOpen {
		return errors.New("não é possível remover pagamentos de uma comanda fechada")
	}

	for i, payment := range c.Payments {
		if payment.ID == paymentID {
			c.TotalRecebido -= payment.ValorRecebido
			c.Payments = append(c.Payments[:i], c.Payments[i+1:]...)
			c.CalculateBalance()
			return nil
		}
	}

	return errors.New("pagamento não encontrado")
}

// RecalculateTotals recalcula os totais da comanda
func (c *Command) RecalculateTotals() {
	c.Subtotal = 0
	for _, item := range c.Items {
		c.Subtotal += item.PrecoFinal
	}

	c.Total = c.Subtotal - c.Desconto
	if c.Total < 0 {
		c.Total = 0
	}

	c.CalculateBalance()
	c.AtualizadoEm = time.Now()
}

// CalculateBalance calcula troco ou saldo devedor
func (c *Command) CalculateBalance() {
	diferenca := c.TotalRecebido - c.Total

	if diferenca > 0 {
		// Cliente pagou a mais -> tem troco
		if c.DeixarTrocoGorjeta {
			c.Troco = 0
			c.SaldoDevedor = 0
		} else {
			c.Troco = diferenca
			c.SaldoDevedor = 0
		}
	} else if diferenca < 0 {
		// Cliente pagou a menos -> tem saldo devedor
		c.Troco = 0
		c.SaldoDevedor = -diferenca
	} else {
		// Pagamento exato
		c.Troco = 0
		c.SaldoDevedor = 0
	}
}

// Helpers
func formatMoney(value float64) string {
	return string(rune(int(value*100) / 100))
}
